### DISCUSSION TITLE
Smart Charging Suspension and etc.

### DISCUSSION THREAD

**Discussion by NP-chaonay** [linux-surface/linux-surface#986] [General]
This is topic for Smart Charging for example:
- Getting if Smart Charging can be request for suspension (for exmaple, on my first portion of time using SLS, it cannot press the suspension button)
- Request suspension of Smart Charging
  - when Smart Charging is not in active
  - when it in active
- Get if Smart Charging is in active
- Get is Smart Charging Suspension is already in active
- Battery driver problem when Smart Charging or Smart Charging Suspension is active

Welcomed for any device model supported Smart Charging

Device configuration can be Battery Limit Mode enabled or disabled

But we assume by deafult that it disabled unless explicitly mentioned. 

Recommended to test without BLM enabled.

Reverse Engineering Guide: https://github.com/linux-surface/surface-aggregator-module/wiki/Development

**Comment by NP-chaonay** [comment]
For device: Surface Laptop Studio

Progress: waiting Smart Charging becoming active, before doing RevEng

Problem Found:
- When SC is enabled, battery reported as 100% and 100% of current capacity  instead of 80 % and 80% of current capacity

**Reply by NP-chaonay** [reply]
[GetIsSCActive_ViaSDMC.zip](https://github.com/linux-surface/linux-surface/files/9997178/GetIsSCActive_ViaSDMC.zip)
This is file getting from IRPMon when executed command "sdmc batterypropget IsSmartChargingEnabled BooleanType"

under the condition:
- SC is not active
- command run without error returned by SDMC
- SC Suspension is not active
- the command returned False boolean
- Only TouchScreen and KBD is use for interact the UI to reduce log

PS: this command is show the active state, not enable state btw.

**Reply by qzed** [reply]
So there seems to be some new stuff, but I'm not entirely sure if it's related to the smart charging stuff:

```json
{"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":239,"rqid_hi":75,"cid":65},"payload":[],"time":"13/11/2022 12:31:37"}
{"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":239,"rqid_hi":75,"cid":65},"payload":[0,0],"time":"13/11/2022 12:31:37"}
```
```json
{"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":240,"rqid_hi":75,"cid":66},"payload":[],"time":"13/11/2022 12:31:37"}
{"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":240,"rqid_hi":75,"cid":66},"payload":[0,0,0,0],"time":"13/11/2022 12:31:37"}

```
These are two requests from the OS to the EC, each with a response. The response payload is always zeros, in both logs provided in the zip. Also, the target category is TC = 1, so SAM directly. For the battery protection I'd expect TC = 2.

It's possible that one of these responses indicates `IsSmartChargingEnabled=0/false`. I guess we'll need a log where we know `IsSmartChargingEnabled=1/true` to confirm.

Apart from that, there are some temperature sensor read-outs and HID input events. There is nothing at all related to the battery subsystem (TC = 2), neither known nor new commands/requests/events.

**Reply by NP-chaonay** [reply]
@qzed so does 4lines is come from either log not both? since time is the same

**Reply by NP-chaonay** [reply]
btw I try these, no output
![image](https://user-images.githubusercontent.com/17851493/201617920-235765dd-ce00-4671-bd8f-7d8a9d67937b.png)

**Reply by qzed** [reply]
Those lines come from the second log, I think. But they are present in both.

> btw I try these, no output

You'll need to set the last byte to 1. Otherwise the command will be interpreted as having no response and it won't wait for one. Also it's decimal above and not hex. So try
```
./ctrl.py 0x01 0x01 0x41 0x00 0x01
```
and
```
./ctrl.py 0x01 0x01 0x42 0x00 0x01
```

**Reply by NP-chaonay** [reply]
still no output 
![image](https://user-images.githubusercontent.com/17851493/201643004-1f6bb824-64ce-428f-afd1-ccf40762d8e6.png)

**Reply by NP-chaonay** [reply]
even ./events.py listen 0x1
and same for 0x2

then run ctrl.py

dont output

**Reply by qzed** [reply]
Ah, I forgot to add the `request`... again. Can you try

```
./ctrl.py request 0x01 0x01 0x41 0x00 0x01
```

```
./ctrl.py request 0x01 0x01 0x42 0x00 0x01
```

**Reply by NP-chaonay** [reply]
> Ah, I forgot to add the `request`... again. Can you try
> 
> ```
> ./ctrl.py request 0x01 0x01 0x41 0x00 0x01
> ```
> 
> ```
> ./ctrl.py request 0x01 0x01 0x42 0x00 0x01
> ```

1) 00 00
2) 00 00 00 00

**Reply by qzed** [reply]
Okay, so same as on Windows. This is all with smart charging disabled, right? Can you try to check those two commands once it's enabled again?

**Reply by NP-chaonay** [reply]
> Okay, so same as on Windows. This is all with smart charging disabled, right? Can you try to check those two commands once it's enabled again?

ok so let my SLS going to smart charging, I have put it 100% all the time, and using all day,

**Reply by NP-chaonay** [reply]
@qzed 

[testWhenSmartChargeActive.zip](https://github.com/linux-surface/linux-surface/files/10027553/testWhenSmartChargeActive.zip)

will try on ubuntu later (will update inplace)


On Ubuntu: this is output
```
root@np-chaonay-l1l:/tmp/tmp.mBP5ZKUDmE/surface-aggregator-module/scripts/ssam# ./ctrl.py request 0x01 0x01 0x41 0x00 0x01
e0 00
root@np-chaonay-l1l:/tmp/tmp.mBP5ZKUDmE/surface-aggregator-module/scripts/ssam# ./ctrl.py request 0x01 0x01 0x42 0x00 0x01
00 00 00 00
```

btw I found problem driver reading the wrong  value of capacity when SC is enabled

**Reply by NP-chaonay** [reply]
[PressChargeTo100InSurfaceApp.zip](https://github.com/linux-surface/linux-surface/files/10027679/PressChargeTo100InSurfaceApp.zip)
This is when I press "charge to 100%" in Surface app.

PS: SUrface cannot undo this request, so there will be no "limit to 80%" button.

**Reply by qzed** [reply]
In the first "charge to 100%" log is:
```json
{"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":20,"rqid_hi":6,"cid":67},"payload":[],"time":"17/11/2022 04:09:13"}
{"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":20,"rqid_hi":6,"cid":67},"payload":[0,0,0,0],"time":"17/11/2022 04:09:13"}
```
in the second one this command is also present, but has a different return value:
```json
{"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":21,"rqid_hi":9,"cid":67},"payload":[],"time":"17/11/2022 04:10:36"}
{"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":21,"rqid_hi":9,"cid":67},"payload":[2,0,0,0],"time":"17/11/2022 04:10:36"}
```

In addition, there is also an additional event in the first log:
```json
{"cmd":{"type":128,"tc":2,"sid":1,"tid":0,"iid":1,"rqid_lo":2,"rqid_hi":0,"cid":24},"payload":[0,0],"time":"17/11/2022 04:09:14"}
```
Based on reverse-engineering, we know that that is a "protection status changed" event. Which seems to fit as this is sent shortly after the CID 67 command.

There are also CID 65/66 commands in the logs. In particular:
- smart-charging active:
  ```json
  {"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":198,"rqid_hi":69,"cid":65},"payload":[],"time":"17/11/2022 03:39:33"}
  {"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":198,"rqid_hi":69,"cid":65},"payload":[96,1],"time":"17/11/2022 03:39:33"}
  {"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":199,"rqid_hi":69,"cid":66},"payload":[],"time":"17/11/2022 03:39:33"}
  {"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":199,"rqid_hi":69,"cid":66},"payload":[0,0,0,0],"time":"17/11/2022 03:39:33"}
  ```
- smart-charging inactive:
  ```json
  {"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":36,"rqid_hi":6,"cid":65},"payload":[],"time":"17/11/2022 04:09:15"}
  {"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":36,"rqid_hi":6,"cid":65},"payload":[0,0],"time":"17/11/2022 04:09:15"}
  {"cmd":{"type":128,"tc":1,"sid":0,"tid":1,"iid":0,"rqid_lo":38,"rqid_hi":6,"cid":66},"payload":[],"time":"17/11/2022 04:09:15"}
  {"cmd":{"type":128,"tc":1,"sid":1,"tid":0,"iid":0,"rqid_lo":38,"rqid_hi":6,"cid":66},"payload":[0,0,0,0],"time":"17/11/2022 04:09:15"}
  ```

Interestingly the values for "active" are different than what you got on Linux.

Once smart-charging is active again, you could try to run
```sh
./ctrl.py request 0x01 0x01 0x43 0x00 0x01
```
I believe this should force charging to 100 / disable smart-charging.

I wonder if there is a 68/0x44  or 0x40 command. That would be my best guess for force-enabling smart charging. So maybe you could also try
```sh
./ctrl.py request 0x01 0x01 0x44 0x00 0x01
```
and
```sh
./ctrl.py request 0x01 0x01 0x40 0x00 0x01
```
and then run the 0x41/0x42 commands again.

**Reply by NP-chaonay** [reply]
@qzed 

Note: There is some reason that make me desired opt to skip some test/testing-condition, the  ex reason: reduce SSD wear from hibernate or battery care.
 - However if you need some need result of specific test/condition, you can explicit tell me btw.

<details><summary>Test Segment #1</summary>

```
#  Init test
cd /home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam
modprobe surface_aggregator_cdev

# Enable SC
function enableSc() {
	./ctrl.py request 0x01 0x01 0x44 0x00 0x01
	./ctrl.py request 0x01 0x01 0x40 0x00 0x01
}
# Check SC
function checkSc() {
	./ctrl.py request 0x01 0x01 0x41 0x00 0x01
	./ctrl.py request 0x01 0x01 0x42 0x00 0x01
}
# Disable SC
function disableSc() {
	./ctrl.py request 0x01 0x01 0x43 0x00 0x01
}

## Test 1 (BLM enabled, the SC state given by sdmc is False)
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
a0 02
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# enableSc
01 00 00 00
01 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc
a0 02
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# disableSc
02 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
a0 02
00 00 00 00

## Test 2 (BLM disabled, tested from test 1 imediently, not know the state of SC in Windows)
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc
00 00
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# enableSc 
01 00 00 00
03 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
00 00
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# disableSc 
02 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
00 00
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# 

# Test 3
# have try 2x disableSc btw before doing this
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# enableSc; sleep 5; echo "SEP"; enableSc 
01 00 00 00
01 00 00 00
SEP
01 00 00 00
01 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# 

root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# disableSc; sleep 5; echo "SEP"; disableSc 
02 00 00 00
SEP
02 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# 
``` 

</details>

<details><summary>Test Segment #2</summary>

- even battery is >=80%, force enable could not be used
output:
```
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# enableSc 
01 00 00 00
01 00 00 00
# after 10sec
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
00 00
00 00 00 00
# even power_now file still reported like pc in charging (>=30000mW, which is definitly in charging)
```

</details>

New test in new comment

**Reply by qzed** [reply]
Hmm, I can't really make sense of the value returned by `0x41` command. That seems to change depending on something, including smart-charging status. 0x42 on the other hand seems to always return zero... So no idea what that command does.

`0x43`/disable seems to return `2` when it's already been disabled. Maybe the return value is some kind of error code. We'll probably have to dig into the Windows driver/app binary and figure that that from there.

**Reply by NP-chaonay** [reply]
@qzed the result came!

after SC agined auto enabled I boot to Linux and do this (cdev is already modprobed before below command)

and it worked

```
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
e0 00
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# disableSc 
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
00 00
00 00 00 00
```

**Reply by NP-chaonay** [reply]
> and it worked

when reboot back to Windows, Surface app and battery status also show that it is charging >80%

**Reply by NP-chaonay** [reply]
ok so to sump up what I test

check SC can be done through `./ctrl.py request 0x01 0x01 0x41 0x00 0x01` if it outputs:
- 00 00 : normal
- e0 00 : SC active
- 20 03 : BLM enabled (I have just done new test after the recent SC disable test, by after SC disable,I reboot and enable BLM, the battery is >=80% at that moment)
- a0 02 : BLM enabled (see above test)
so I conclude that for SLS

to disable, using `./ctrl.py request 0x01 0x01 0x43 0x00 0x01`

gives `00 00 00 00` means just disable in non-BLM

gives `02 00 00 00` means already disabled

I also haved try another  than 0x43 (such as 0x40 0x3f 0x44 ... 0x48 0x49 (0x49 gives ETIMEDOUT) ) none of them work for re-enable (it seem no way to set enable idk)

**Reply by NP-chaonay** [reply]
this is log when BLM disabled and batter ystill >=80%, SC have been disabled by user.

```
function enableSc() {
        ./ctrl.py request 0x01 0x01 0x3f 0x00 0x01
        ./ctrl.py request 0x01 0x01 0x40 0x00 0x01
        ./ctrl.py request 0x01 0x01 0x43 0x01 0x01
        ./ctrl.py request 0x01 0x01 0x43 0x02 0x01
        ./ctrl.py request 0x01 0x01 0x44 0x00 0x01
        ./ctrl.py request 0x01 0x01 0x45 0x00 0x01
        ./ctrl.py request 0x01 0x01 0x46 0x00 0x01
        ./ctrl.py request 0x01 0x01 0x47 0x00 0x01
        ./ctrl.py request 0x01 0x01 0x48 0x00 0x01
}
....
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# 
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# 
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# checkSc 
00 00
00 00 00 00
root@np-chaonay-l1l:/home/np-chaonay/Desktop/Temp/ComDebug/surface-aggregator-module/scripts/ssam# enableSc 
00
02 00 00 00
02 00 00 00
02 00 00 00
01 00 00 00
ff ff ff ff
01
00 00 00 00
00
```

**Reply by qzed** [reply]
I suspect that the value returned by `./ctrl.py request 0x01 0x01 0x41 0x00 0x01` is a bit field:
- `00 00 = 0x0000 = 0000 0000 0000 0000` (normal)
- `e0 00 = 0x00e0 = 0000 0000 1110 0000` (SC)
- `20 03 = 0x0320 = 0000 0011 0010 0000` (BLM 1)
- `a0 02 = 0x02a0 = 0000 0010 1010 0000` (BLM 2)

So we maybe have something like
- `?? ?? = 0x???? = ???? ??L? ?S?? ????`

where
- `L` is the bit for battery limiter enabled and
- `S` is the bit for smart-charging enabled

Maybe we can validate this by some reverse engineering? Any idea what we would have to look at specifically (i.e. which `.dll` or `.exe`)?

**Reply by NP-chaonay** [reply]
> Maybe we can validate this by some reverse engineering? Any idea what we would have to look at specifically (i.e. which .dll or .exe)?

it seem that there is standard API in Win11 for battery driver to send to let Windows know if smart charging mode is enabled

so I think if we want to do do further RevEng on file inspection method, we might inspect trhough driver file (but seem it diffcult so much) or maybe Surface app executable file.

other than this, i have no idea.

**Reply by qzed** [reply]
Okay, so maybe the battery driver. The Surface App would probably be a good starting point but unfortunately I don't know much about how Windows Apps from their app-store work other than that they aren't regular `.exe` files.

**Comment by NP-chaonay** [comment]
Ok for non-SLS device user.

Checking SC state and BLM state and Disable SC is now work on SLS Linux-Surface!

you can have try on my info
